src/canary.rs

```diff
@@ -186,3 +185,88 @@ impl Canary {
         }
     }
 }
+
+/// Write [`CANARY_VALUE`] to the stack.
+fn paint_stack(core: &mut Core, start: u32, mut end: u32) -> Result<(), probe_rs::Error> {
+    // make sure stack_end is properly aligned
+    if end % 4 != 0 {
+        end += end % 4;
+    }
+
+    // does the subroutine fit inside the stack?
+    let stack_size = (end - start) as usize;
+    assert!(
+        SUBROUTINE_LENGTH < stack_size,
+        "subroutine doesn't fit inside stack"
+    );
+
+    // write subroutine to RAM
+    // NOTE: add `SUBROUTINE_LENGTH` to `start`, to avoid the subroutine overwriting itself
+    core.write_8(start, &subroutine(start + SUBROUTINE_LENGTH as u32, end))?;
```

Comment:
If we're 4-byte aligned, I guess we could use `write_32` here? Not sure it makes much difference though.

Reply:
I don't think this works so well. It would require us to make `fn subroutine` return a `[u32; _]`, which would require is to combine two instructions (4 bytes) into one, which makes it less readable imo.

Or what do you think?

```diff
-    [
-        0x03, 0x48, // ldr
-        0x04, 0x49, // ldr
-        0x04, 0x4a, // ldr
-        0x81, 0x42, // cmp
-        0x01, 0xD0, // beq.n
-        0x04, 0xC0, // stmia
-        0xFB, 0xE7, // b.n
-        0x00, 0xBE, // bkpt
-        s1, s2, s3, s4, // start address
-        e1, e2, e3, e4, // end address
-        CV, CV, CV, CV, // canary value
-    ]
+    [
+        0x03480449, // ldr, ldr
+        0x044A8142, // ldr, cmp
+        0x01D004C0, // beq.n, stmia
+        0xFBE700BE, // b.n, bkpt
+        start, // start address
+        end, // end address
+        canary, // canary value
+    ]
```